File: haskell/Exercises.hs
module Exercises
    ( change,
    firstThenApply,
    powers,
    meaningfulLineCount,
    Shape(..),
    volume,
    surfaceArea,
    BST(Empty, Node),
    size,
    inorder,
    insert,
    contains
    -- empty
    --toString
      -- put the proper exports here
    ) where

import qualified Data.Map as Map
import Data.Text (pack, unpack, replace)
import Data.List(isPrefixOf, find)
import Data.Char(isSpace)

change :: Integer -> Either String (Map.Map Integer Integer)
change amount
    | amount < 0 = Left "amount cannot be negative"
    | otherwise = Right $ changeHelper [25, 10, 5, 1] amount Map.empty
        where
          changeHelper [] remaining counts = counts
          changeHelper (d:ds) remaining counts =
            changeHelper ds newRemaining newCounts
              where
                (count, newRemaining) = remaining `divMod` d
                newCounts = Map.insert d count counts

-- Write your first then apply function here
firstThenApply :: [a] -> (a -> Bool) -> (a -> b) -> Maybe b
firstThenApply xs predicate func = fmap func (find predicate xs)

-- Write your infinite powers generator here
powers :: Integer -> [Integer]
powers base = map (base^) [0..]

-- Write your line count function here
meaningfulLineCount :: FilePath -> IO Int
meaningfulLineCount path = do
    contents <- readFile path
    return $ length $ filter meaningfulLine $ lines contents
    where
      meaningfulLine line = not (all isSpace line) && not ("#" `isPrefixOf` (dropWhile isSpace line))

-- Write your shape data type here
data Shape = Sphere Double | Box Double Double Double 
    deriving (Eq, Show)

volume :: Shape -> Double
volume (Sphere r) = (4 / 3) * pi * r^3
volume (Box w h d) = w * h * d

surfaceArea :: Shape -> Double
surfaceArea (Sphere r) = 4 * pi * r^2
surfaceArea (Box w h d) =  2 * (w * h + w * d + h * d)

-- Write your binary search tree algebraic type here


data BST a = Empty | Node a (BST a) (BST a) 

size :: BST a -> Int
size Empty = 0
size (Node _ left right) = 1 + size left + size right

contains :: (Ord a) => a -> BST a -> Bool
contains _ Empty = False
contains item (Node target left right)
    | item < target = contains item left
    | item > target = contains item right
    | otherwise = True

insert :: (Ord a) => a -> BST a -> BST a
insert item Empty = Node item Empty Empty
insert item (Node target left right)
    | item < target = Node target (insert item left) right
    | item > target = Node target left (insert item right)
    | otherwise = Node target left right  

inorder :: BST a -> [a]
inorder Empty = []
inorder (Node item left right) = inorder left ++ [item] ++ inorder right

toString :: (Show a) => BST a -> String
toString Empty = "()"
toString (Node item left right) = "(" ++ toString left ++ " " ++ show item ++ " " ++ toString right ++ ")"

removeExtraParen :: String -> String
removeExtraParen f = unpack (replace (pack "()") (pack "") (pack f))
instance (Show a) => Show (BST a) where
    show :: Show a => BST a -> String
    show Empty = "()"
    show (Node value left right) = removeExtraParen ("(" ++ show left ++ show value ++ show right ++ ")")

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

(* Write your first then apply function here *)
let first_then_apply array predicate transformer = 
  match List.find_opt predicate array with
  | None -> None
  | Some x -> transformer x;;

(* Write your powers generator here *)
let powers_generator base = 
  let rec generate_from power () =
    Seq.Cons(power, generate_from (power * base))
  in
  generate_from 1;;

(* Write your line count function here *)
let meaningful_line_count file_path =
  let count_lines ic =
    let rec count acc =
      match In_channel.input_line ic with
      | None -> acc
      | Some line ->
        let trimmed = String.trim line in
        if trimmed = "" || String.get trimmed 0 = '#' then
          count acc
        else
          count (acc + 1)
    in
    count 0
  in
  In_channel.with_open_text file_path (fun ic ->
    Fun.protect
      ~finally:(fun () -> In_channel.close ic)
      (fun () -> count_lines ic)
  )

(* Write your shape type and associated functions here *)
type shape = 
  | Sphere of float
  | Box of float * float * float
  [@@deriving eq, show]

let surface_area = function 
  | Sphere r -> 4.0 *. Float.pi *. r ** 2.0
  | Box (w,h,d) -> 2.0 *. (d *. w +. w *. h +. h *. d)

let volume = function
  | Sphere r -> (4.0 /. 3.0) *. Float.pi *. r ** 3.0
  | Box (w,h,d) -> w *. h *. d

(* Write your binary search tree implementation here *)

type 'a binary_search_tree = 
  | Empty
  | Node of 'a binary_search_tree * 'a * 'a binary_search_tree

  let empty = Empty

  let rec insert target = function
    | Empty -> Node (Empty, target, Empty)
    | Node (left, item, right) as node ->
      if target < item then 
        Node (insert target left, item, right)
      else if target > item then 
        Node (left, item, insert target right)
      else node

  let rec contains target = function
    | Empty -> false
    | Node (left, item, right) ->
      if target < item then 
        contains target left
      else if target > item then 
        contains target right
      else true

  let rec size = function
    | Empty -> 0
    | Node (left, _, right) -> 1 + size left + size right

  let rec inorder = function
    | Empty -> []
    | Node (left, item, right) -> inorder left @ [item] @ inorder right
